src/worker/swirl/runner.rs

```diff
@@ -26,155 +23,153 @@ fn runnable<J: BackgroundJob>(ctx: &J::Context, payload: serde_json::Value) -> a
                     job_registry: self.job_registry.clone(),
-                    sender: sender.clone(),
+                    shutdown_when_queue_empty: self.shutdown_when_queue_empty,
+                    poll_interval: self.poll_interval,
                 };
 
-                self.thread_pool.execute(move || worker.run_next_job())
-            }
+                std::thread::Builder::new()
+                    .name(name.clone())
+                    .spawn(move || {
+                        info_span!("worker", worker.name = %name).in_scope(|| worker.run())
+                    })
+            })
+            .collect::<Result<_, _>>()?;
 
-            pending_messages += jobs_to_queue;
-            match receiver.recv_timeout(self.job_start_timeout)? {
-                Event::Working => pending_messages -= 1,
-                Event::NoJobAvailable => return Ok(()),
-                Event::ErrorLoadingJob(e) => return Err(FetchError::FailedLoadingJob(e)),
-                Event::FailedToAcquireConnection(e) => {
-                    return Err(FetchError::NoDatabaseConnection(e));
-                }
-            }
-        }
+        Ok(RunningRunner { handles })
     }
 
     fn connection(&self) -> Result<DieselPooledConn, PoolError> {
         self.connection_pool.get()
     }
 
-    /// Waits for all running jobs to complete, and returns an error if any
-    /// failed
+    /// Check if any jobs in the queue have failed.
     ///
-    /// This function is intended for use in tests. If any jobs have failed, it
-    /// will return `swirl::JobsFailed` with the number of jobs that failed.
-    ///
-    /// If any other unexpected errors occurred, such as panicked worker threads
-    /// or an error loading the job count from the database, an opaque error
-    /// will be returned.
-    // FIXME: Only public for `src/tests/util/test_app.rs`
+    /// This function is intended for use in tests.
     pub fn check_for_failed_jobs(&self) -> anyhow::Result<()> {
-        self.wait_for_jobs()?;
         let failed_jobs = storage::failed_job_count(&mut *self.connection()?)?;
         if failed_jobs == 0 {
             Ok(())
         } else {
             Err(anyhow!("{failed_jobs} jobs failed"))
         }
     }
+}
 
-    fn wait_for_jobs(&self) -> anyhow::Result<()> {
-        self.thread_pool.join();
-        let panic_count = self.thread_pool.panic_count();
-        if panic_count == 0 {
-            Ok(())
-        } else {
-            Err(anyhow!("{panic_count} threads panicked"))
-        }
+pub struct RunningRunner {
+    handles: Vec<JoinHandle<()>>,
+}
+
+impl RunningRunner {
+    /// Wait for all background workers to shut down.
+    pub fn wait_for_shutdown(self) {
+        self.handles.into_iter().for_each(|handle| {
+            if let Err(error) = handle.join() {
+                let error = try_to_extract_panic_info(&error);
+                warn!(%error, "Background worker thread panicked");
+            }
+        })
     }
 }
```

Comment:
I can definitely see how this would be adapted to a set of tasks (\*cough\* `join_all` \*cough\*). :+1: